exporter/elasticsearchexporter/pdata_serializer.go

```diff
@@ -0,0 +1,372 @@
+		writeStringFieldSkipDefault(v, "trace_id", spanLink.TraceID().String())
+		writeStringFieldSkipDefault(v, "span_id", spanLink.SpanID().String())
+		writeStringFieldSkipDefault(v, "trace_state", spanLink.TraceState().AsRaw())
+		writeAttributes(v, spanLink.Attributes(), false)
+		writeIntFieldSkipDefault(v, "dropped_attributes_count", int64(spanLink.DroppedAttributesCount()))
+		_ = v.OnObjectFinished()
+	}
+	_ = v.OnArrayFinished()
+}
+
+func serializeLog(resource pcommon.Resource, resourceSchemaURL string, scope pcommon.InstrumentationScope, scopeSchemaURL string, record plog.LogRecord) ([]byte, error) {
+	var buf bytes.Buffer
+
+	v := json.NewVisitor(&buf)
+	// Enable ExplicitRadixPoint such that 1.0 is encoded as 1.0 instead of 1.
+	// This is required to generate the correct dynamic mapping in ES.
+	v.SetExplicitRadixPoint(true)
+	_ = v.OnObjectStart(-1, structform.AnyType)
+	docTimeStamp := record.Timestamp()
+	if docTimeStamp.AsTime().UnixNano() == 0 {
+		docTimeStamp = record.ObservedTimestamp()
+	}
+	writeTimestampField(v, "@timestamp", docTimeStamp)
+	writeTimestampField(v, "observed_timestamp", record.ObservedTimestamp())
+	writeDataStream(v, record.Attributes())
+	writeStringFieldSkipDefault(v, "severity_text", record.SeverityText())
+	writeIntFieldSkipDefault(v, "severity_number", int64(record.SeverityNumber()))
+	writeTraceIDField(v, record.TraceID())
+	writeSpanIDField(v, "span_id", record.SpanID())
+	writeAttributes(v, record.Attributes(), false)
+	writeIntFieldSkipDefault(v, "dropped_attributes_count", int64(record.DroppedAttributesCount()))
+	if record.EventName() != "" {
+		writeStringFieldSkipDefault(v, "event_name", record.EventName())
+	} else if eventNameAttr, ok := record.Attributes().Get("event.name"); ok && eventNameAttr.Str() != "" {
+		writeStringFieldSkipDefault(v, "event_name", eventNameAttr.Str())
+	}
+	writeResource(v, resource, resourceSchemaURL, false)
+	writeScope(v, scope, scopeSchemaURL, false)
+	writeLogBody(v, record)
+	_ = v.OnObjectFinished()
+	return buf.Bytes(), nil
+}
+
+func writeDataStream(v *json.Visitor, attributes pcommon.Map) {
+	_ = v.OnKey("data_stream")
+	_ = v.OnObjectStart(-1, structform.AnyType)
+	attributes.Range(func(k string, val pcommon.Value) bool {
+		if strings.HasPrefix(k, "data_stream.") && val.Type() == pcommon.ValueTypeStr {
+			writeStringFieldSkipDefault(v, k[12:], val.Str())
+		}
+		return true
+	})
+
+	_ = v.OnObjectFinished()
+}
+
+func writeLogBody(v *json.Visitor, record plog.LogRecord) {
+	if record.Body().Type() == pcommon.ValueTypeEmpty {
+		return
+	}
+	_ = v.OnKey("body")
+	_ = v.OnObjectStart(-1, structform.AnyType)
+
+	// Determine if this log record is an event, as they are mapped differently
+	// https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md
+	var bodyType string
+	if _, hasEventNameAttribute := record.Attributes().Get("event.name"); hasEventNameAttribute || record.EventName() != "" {
```

Comment:
We already do this check in the caller, and in a more efficient way, so should we make it a bool parameter?

Reply:
Done in 7ba257562886184311eee8058ccff4a1c8f485d0